### ISSUE TITLE
reportOverlappingOverload disagrees with observed behavior

### ISSUE THREAD

**Issue by CGamesPlay** [microsoft/pyright#8465]
This is a continuation of #8389. In that thread, you worked with me through a few iterations to highlight some mistakes in my reasoning, but I am presenting a simple case now that I hope we can both agree is a bug.

When you run the following code, Pyright says "Overload 2 for "decorator" will never be used because its parameters overlap overload 1", however we can clearly see that overload 2 was selected when the argument was not compatible with overload 1. The behavior is correct, the warning is incorrect.

Code sample in [pyright playground](https://pyright-play.net/?enableExperimentalFeatures=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFAMICGANtZQEbUCmZYAbkyNWJQCZkgmHGgH14CJgChpAAXadufSbybAoKgMa5KMXAApgALgo06jJgG0LqGAF0yAZxghbASigBaAHyYUMYwB0QZJyHFw8vMqq6kxaIDq4FgDKtgbGVLQMzFYpjs5unj7Aiv5QQQFRaprauiBphMTu3g1whpJQHVCCMACuIChQwNKVUPLhfCIAjHoa9OmmWZbWfvZQTi6ugcEqamOKvCIATMmps-OZ5jmr625bFdI7sJQA1kwOImD0AFaxMHoAHsZPj8NDAmj51ncRjAXm8ROsAcZ1uC1s4odI9hEpnoYa93sDfq5JJiJtNcXDkcSwvsjjjYfjvoSqQosYc6Xj4c4iYJhNQxIgmHpqvFauy4QTQa53FAAMRQADy1IiUEmayYzFBTF4UD0Nm5QiYonEguFCTq5PeyOlcsVLL4UEOao1MC1OuKPDBkiAA)

```python
@overload
def decorator(f: Callable[[int], str]) -> int: ...
@overload
def decorator[S](f: Callable[[S], str]) -> float: ...
def decorator(f: Any) -> Any:
    return f

def takes_object(x: object) -> str: ...
def takes_str(x: str) -> str: ...

reveal_type(decorator(takes_object))  # Overload 1 selected (int)
reveal_type(decorator(takes_str))  # Overload 2 selected (float)
```

---

While I've got your attention, I wanted to take a second to deliver my respects for your amazing work on this project. I've now submitted 2 issues I've found in the course of my work with the tool, and in both cases you've been extremely helpful, friendly, and quick to address the feedback and even correct the issues. It's extremely impressive. Thank you for being so awesome!

**Comment by randolf-scholz** [comment]
There are 2 issues here:

1. The Type Variable in the second overload is kind of meaningless, because it only appears once. You should use a concrete type instead. `pyright` actually complains about this if we simplify a bit: Code sample in [pyright playground](https://pyright-play.net/?enableExperimentalFeatures=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFGAG4CmIANmAIYAmAUCwAKU33MtNXAowMGAAUwAFyYUMAJRQAtAD5pMKQDpNHLnUat%2Bg4WADaAZQC64qafnKhPNVE3q%2BAoSKuFitlUThSQKhgAVxAUISA)

```python
from typing import Any, overload

@overload
def foo(f: int) -> int: ...
@overload
def foo[S](f: S) -> float: ...
def foo(f: Any) -> Any: return f
```

>\>\>\> TypeVar "S" appears only once in generic function signature  Use "object" instead  (reportInvalidTypeVarUse)

2. The second issue is you are overlooking the contra-variance of the argument type. Let's imagine we substituted `Callable[[object], str]` in the second overload. But notice that `Callable[[object], str] <: Callable[[int], str]`, that is the Callable with the wider argument type, is indeed a subtype of the callable with narrower argument type. (Convince yourself that this is true!)

What you seemingly want to express is:

> If the function `f` accepts `int`, then return an `int`. If it doesn't then return `float`.

Therefore, the argument type in the second overload needs to be *narrower* than `int`. I think the (currently) correct way to express this is via 

```python
@overload
def decorator(f: Callable[[int], str]) -> int: ...
@overload
def decorator(f: Callable[[Never], str]) -> float: ... 
def decorator(f: Any) -> Any:
    return f
```

Code sample in [pyright playground](https://pyright-play.net/?enableExperimentalFeatures=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFAMICGANtZQEbUCmZYAbkyNWJQCZkgmHGgH14CFlAByQzgCg5AAXadufObybAomgMa5KMXAApgALgo06jJgG1bqGAF0yAZxggnASigBaAHyYKDAWAHThSipcPLwaWjpM%2BiCGJuaWtAzM9jIcnm4e3n6BwGohUOGhcdp6BkYgphZEcD4BhMRmclBdUIIwAK4gKFDAClVQUWq8IgCMxrr0FlQZNvaOLlDunl5hEZraEzEiAEy2AMpOcwvp1lln65veO5UKe7CUANZMriJg9ABWiRgxgAHhZfgDdDAWoFNk8xjAPl8RJsQRZNtCNh44QoDnwZsYEZ9vuDAV45LiprNCUj0eTcpNjgTEcT-qS6apDkcmUTkR4yYJhNQxIgmMYask6tykSTIV4fFAAMRQADy9JiUGmGyYzEhTF4UGMjn5slE4lF4pS9Wp33R8qVqo5fCgRy1OpgeoNJR4UKAA)

**Comment by CGamesPlay** [comment]
Hello, thanks for chiming in with the detailed explanation. Substituting the useless generic for `object` and `Never` definitely helps me understand what's going on. So I'll agree the my overloads in this example are not the correct ones for me to use.

Do you have any comment about pyright being self-inconsistent?

> Overload 2 for "decorator" *will never be used* because its parameters overlap overload 1
> Type of "decorator(takes_str)" is "float"

**Comment by randolf-scholz** [comment]
I don't know the internals of `pyright`, so I can only guess, but I think what happens is that when the overloads are compared it substitutes the TypeVar `S` with its upper bound (`Any` or the internal `Unknown`) and then compares.

For instance, if you put an upper bound of `str` the error goes away: Code sample in [pyright playground](https://pyright-play.net/?enableExperimentalFeatures=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFAMICGANtZQEbUCmZYAbkyNWJQCZkgmHGgH14CJgChpAAXadufSbybAoKgMa5KMXAApgALgo06jJgG0LqGAF0yAZxghbASigBaAHyYUMYwB0QZJyHFw8vMqq6kxaIDq4FgDKxk4uBsZUtAzMVkn2UGlunj7Aiv5QQQFRaprauiAZhMTu3s1whpJQ3VCCMACuIChQwNI1UPLhfCIAjHoa9JmmOZbWfgVFroHBKmqTirwiAEzJtvOLJtnmeRvObtvV0ruwlADWTA4iYPQAVrEwegAHsZvn8NDBWj40g9xjA3h8RGkgalnJDCs4YdJ9hFZno4e9PqD-q5JNjpnN8Qi0iSyYcjnj4YTfsTSWEDscGQTEajJIJhNQxIgmHo6vEGpyEUTwa53FAAMRQADybIiUBmhSYzHBTF4UD0NhJfKYonEwtFCUalM%2B1NlCuVClVRw1WpgOr1ZR4EMkQA)

**Comment by CGamesPlay** [comment]
That seems like a useful insight! For covariant generic types, the type checker can demonstrate that an overload will never be used if the upper bound is a subtype of one of the other overloads, because that overload would already have matched.

But for contravariant types, to prove that an overload can never be used, the type checker needs to prove the nonexistence of any lower bound that the generic could use.

**Comment by randolf-scholz** [comment]
I'd recommend that you post a MWE and what you are actually trying to do over here: https://github.com/python/typing/discussions

**Comment by erictraut** [comment]
I agree this is a bug. I'll investigate the underlying cause.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.373.
